packages/graphcache-rn-async-storage/src/makeAsyncStorage.ts

```diff
@@ -0,0 +1,114 @@
+import { StorageAdapter } from '@urql/exchange-graphcache';
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import NetInfo from '@react-native-community/netinfo';
+
+export type StorageOptions = {
+  dataKey?: string;
+  metadataKey?: string;
+  maxAge?: number; // Number of days
+};
+
+let disconnect;
+
+export const makeAsyncStorage: (ops?: StorageOptions) => StorageAdapter = ({
+  dataKey = 'graphcache-data',
+  metadataKey = 'graphcache-metadata',
+  maxAge = 7,
+} = {}) => {
+  const todayDayStamp = Math.floor(
+    new Date().valueOf() / (1000 * 60 * 60 * 24)
+  );
+  const todayBatch = {};
+  const prefixCheck = new RegExp(`^${dataKey}_(\\d+)$`);
+
+  return {
+    /**
+     * On initial read, pull storage keys and see which match our storage key pattern.
+     * Any record that is expired should be removed from AsyncStorage.
+     * The rest of the batches should be merged into a cache in chrono order.
+     * Also hydrate todayBatch if we find it.
+     */
+    readData: async () => {
+      const cache = {};
+
+      try {
+        const persistedDayStamps = (await AsyncStorage.getAllKeys())
+          .reduce<number[]>((filtered, curr) => {
+            const match = curr.match(prefixCheck);
+            const ts = match ? match[1] : undefined;
+
+            if (ts) {
+              filtered.push(Number(ts));
+            }
+            return filtered;
+          }, [])
+          .sort();
+
+        // eslint-disable-next-line es5/no-for-of
+        for (const dayStamp of persistedDayStamps) {
+          // Discard
+          if (todayDayStamp - dayStamp > maxAge) {
+            await AsyncStorage.removeItem(`${dataKey}_${dayStamp}`);
+          }
+          // Parse batch and merge in.
+          else {
+            try {
+              const data = await AsyncStorage.getItem(`${dataKey}_${dayStamp}`);
+              if (data) {
+                const parsedData = JSON.parse(data);
+                Object.assign(cache, parsedData);
+
+                // We found today's batch, let's hydrate that while we're at it.
+                if (dayStamp === todayDayStamp) {
+                  Object.assign(todayBatch, parsedData);
+                }
+              }
+            } catch (_err) {}
+          }
+        }
+      } catch (_err) {}
+
+      return cache;
+    },
+
+    writeData: async delta => {
+      try {
+        Object.assign(todayBatch, delta);
+        await AsyncStorage.setItem(
+          `${dataKey}_${todayDayStamp}`,
+          JSON.stringify(todayBatch)
```

Comment:
I think all these `JSON.stringify` calls hide a very inconvenient bug 😅 The `SerializedEntries` contain `undefined` values which must be serialised or treated in a special manner: https://github.com/FormidableLabs/urql/blob/63eec7eb8b6efeb49c0bb27249dec58d12121d2a/exchanges/graphcache/src/types.ts#L208-L210

In the original implementation, we didn't actually "merge" them to overwrite old values, as we're dealing with batches that are deserialized in order. But the important point is that a key may be set to `undefined` to denote that it has been deleted or garbage collected in Graphcache. It's possible that failing this may lead to some issues. But I'm not entirely sure. https://github.com/FormidableLabs/urql/blob/63eec7eb8b6efeb49c0bb27249dec58d12121d2a/exchanges/graphcache/src/store/data.ts#L623

This is how we're solving it there, I believe: https://github.com/FormidableLabs/urql/blob/63eec7eb8b6efeb49c0bb27249dec58d12121d2a/exchanges/graphcache/src/default-storage/index.ts#L66-L67

Reply:
This should be addressed with this commit: https://github.com/FormidableLabs/urql/pull/1949/commits/0f4696c262321795ead32fa4a86e2cb4677fde12